firedrake/supermeshing.py

```diff
@@ -0,0 +1,432 @@
+            xi = PETSc.Vec().createWithArray(xa, size=sizes[1], comm=x.comm)
+            yi = PETSc.Vec().createWithArray(ya, size=sizes[0], comm=y.comm)
+            self.mat.mult(xi, yi)
+            y.array[start::stride] = yi.array_r
+
+
+def assemble_mixed_mass_matrix(V_A, V_B):
+    """
+    Construct the mixed mass matrix of two function spaces,
+    using the TrialFunction from V_A and the TestFunction
+    from V_B.
+    """
+
+    if len(V_A) > 1 or len(V_B) > 1:
+        raise NotImplementedError("Sorry, only implemented for non-mixed spaces")
+
+    if V_A.ufl_element().mapping() != "identity" or V_B.ufl_element().mapping() != "identity":
+        msg = """
+Sorry, only implemented for affine maps for now. To do non-affine, we'd need to
+import much more of the assembly engine of UFL/TSFC/etc to do the assembly on
+each supermesh cell.
+"""
+        raise NotImplementedError(msg)
+
+    mesh_A = V_A.mesh()
+    mesh_B = V_B.mesh()
+
+    dim = mesh_A.geometric_dimension()
+    assert dim == mesh_B.geometric_dimension()
+    assert dim == mesh_A.topological_dimension()
+    assert dim == mesh_B.topological_dimension()
+
+    (mh_A, level_A) = get_level(mesh_A)
+    (mh_B, level_B) = get_level(mesh_B)
+
+    if mesh_A is mesh_B:
+        def likely(cell_A):
+            return [cell_A]
+    else:
+        if (mh_A is None or mh_B is None) or (mh_A is not mh_B):
+            msg = """
+Sorry, only implemented for non-nested hierarchies for now. You need to
+call libsupermesh's intersection finder here to compute the likely cell
+coverings that we fetch from the hierarchy.
+"""
+
+            raise NotImplementedError(msg)
+
+        if abs(level_A - level_B) > 1:
+            raise NotImplementedError("Only works for transferring between adjacent levels for now.")
+
+        # What are the cells of B that (probably) intersect with a given cell in A?
+        if level_A > level_B:
+            cell_map = mh_A.fine_to_coarse_cells[level_A]
+
+            def likely(cell_A):
+                return cell_map[cell_A]
+
+        elif level_A < level_B:
+            cell_map = mh_A.coarse_to_fine_cells[level_A]
+
+            def likely(cell_A):
+                return cell_map[cell_A]
+
+    assert V_A.value_size == V_B.value_size
+    orig_value_size = V_A.value_size
+    if V_A.value_size > 1:
+        V_A = firedrake.FunctionSpace(mesh_A, V_A.ufl_element().sub_elements()[0])
+    if V_B.value_size > 1:
+        V_B = firedrake.FunctionSpace(mesh_B, V_B.ufl_element().sub_elements()[0])
+
+    assert V_A.value_size == 1
+    assert V_B.value_size == 1
+
+    # for cell_A in range(mesh_A.num_cells()):
+    #     print("likely(%s) = %s" % (cell_A, likely(cell_A)))
+
+    # Preallocate sparsity pattern for mixed mass matrix from likely() function:
+    # For each cell_A, find dofs_A.
+    #   For each cell_B in likely(cell_B),
+    #     Find dofs_B.
+    #     For dof_B in dofs_B:
+    #         nnz[dof_B] += len(dofs_A)
```

Comment:
Could redo these comments to describe algorithmic steps, rather than pseudo-code?